onnxruntime/core/providers/webgpu/nn/conv.cc

```diff
@@ -19,6 +28,24 @@ Status TransposeKernel(ComputeContext& context, const Tensor* kernel, const Tens
     transposed_kernel_shape_vector[i] = kernel_shape[perm[i]];
   }
   uint32_t output_size = onnxruntime::narrow<uint32_t>(kernel_shape.Size());
+
+  uint32_t dispatch_x = ceil_div(output_size, 64);
+  uint32_t dispatch_y = 1;
+  uint32_t dispatch_z = 1;
+
+  // This temporary workaround addresses a significant performance bottleneck
+  // (10x slower) for the shape (3, 3, 2560, 1280) due to an issue with Intel's
+  // GPU drivers. We manually normalize the dispatch group size to restore
+  // performance.
+  //
+  // TODO: Revert this change once the driver issue is fixed.
+  if (context.AdapterInfo().vendor == std::string_view{"intel"}) {
+    ORT_ENFORCE(rank == static_cast<size_t>(4), "Input tensor must have rank 4.");
+    dispatch_x = ceil_div(transposed_kernel_shape_vector[0] * transposed_kernel_shape_vector[1], 2);
+    dispatch_y = ceil_div(transposed_kernel_shape_vector[2], 4);
+    dispatch_z = ceil_div(transposed_kernel_shape_vector[3], 8);
```

Comment:
I am wondering what kind of pattern of the dispatched workgroups will trigger the driver slow path. Does a dispatch group size of (679, 679, 1) resulting in extremely slow performance on LNL apply for any compute shader's dispatch or only for transpose? If it's not transpose specific, should we put the workaround in a common place for intel device?

Reply:
Thanks for the background. I just want to make sure we don't missing something. I am ok if this workaround won't bring any regression on existed models on Intel that @guschmue monitored.

BTW, you can do an easy test that only include one op in the model, for example, transpose with the issued shape to see whether it triggers the driver issue. If yes, your workaround should at least also apply to the transpose op in https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/webgpu/tensor/transpose.cc#L103. Or maybe both of the places reuse the same DoTranspose interface.

Using the similar method, you can try another op , like matmul , to see if it has the similar issue to verify that whether other ops have the dispatching layout issue.

Reply:
`DoTranspose` is in https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/webgpu/tensor/transpose.cc#L103

Reply:
Can you refactor the code and directly call DoTranspose instead of TransposeKernel and put your workaround into DoTranspose so that all transpose can use this workaround?

> After looking at the source code, the refactor is more complicated than I expected. I'll handle it in a separate PR.

And if this's LNL specific issue, please further constraint it to LNL instead of all intel device.

> Probably PTL as well. It's much safer to check the `vendor` here.